### PR TITLE
Offline Logits-Based Knowledge Distillation

### PR DESCRIPTION
**Pull request by AAnoosheh**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?

This PR adds end-to-end infrastructure for **offline knowledge distillation** using cached teacher log-probabilities. The workflow is split into two separate training passes:

1. **Teacher pass** – run the teacher model and save top-K log-probabilities to disk.
2. **Student pass** – train a student model using a forward KL-divergence loss against the cached teacher log-probs.

---

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by AAnoosheh** [comment]
### New: `megatron/training/distillation/`

**`logits_saver.py`** — `LogitsSaverHooks`

Hook-based utility that captures the teacher model's output logits during the forward pass and writes top-K log-probabilities to disk with minimal overhead:
- Top-K on raw logits first (avoids materializing a full-vocab log-softmax tensor).
- Optional **top-P nucleus filtering** (`--logits-save-top-p`) on top of top-K.
- Output is zstd-compressed tensors packed into **tar archives** to reduce inode count. Configurable batching of multiple iterations per tar (`--logits-save-flush-interval`).
- Compact **17-bit vocab index storage**: lower 16 bits as `uint16`, 17th bit stored separately as `bool`.
- Saves shard per DP-CP rank for parallel read/write.
- Remote storage support (MSC / object store paths).

**`cached_logits_loss.py`** — `LossFuncCallable`

Loss function for the student pass that loads pre-computed teacher top-K log-probs and computes forward KL divergence:
- **Sparse KL**: each TP rank operates only on teacher positions falling in its local vocab shard — no full-vocab dense teacher tensor.
- Configurable blending of KD loss and standard LM loss via `--logits-load-kd-loss-alpha`.
- **Multi-threaded decode pipeline**: tar `IterableDataset` with an internal `ThreadPoolExecutor` parallelises zstd decompression + `torch.load` + index unpacking, removing the single-thread bottleneck.
- **DP-resharding**: supports student runs with a different DP size than the teacher (both up- and down-scaling, as long as one evenly divides the other).
- MSC/object-store prefetching of whole tar shards (`--logits-load-msc-prefetch-depth`).

**`utils_logits.py`** — Shared utilities (tar I/O, remote storage helpers, dataset hash validation, index packing/unpacking, etc.)

---

### Training integration (`megatron/training/training.py`)

- `setup_model_and_optimizer` instantiates and attaches `LogitsSaverHooks` (teacher pass) or `StudentLogitsCapture` hooks (student pass) based on arguments.
- Teacher pass skips the LM loss entirely (replaced with a zero tensor) for speed.
- `--freeze-all-layers` freezes all model parameters (and MoE expert biases) during the teacher inference pass.
- `LogitsSaverHooks.shutdown()` is called after training to flush any buffered data.
- Minor fix: `learning_rate` set to `-1.0` as a sentinel when all layers are frozen, to satisfy existing logging logic.

---

### Checkpointing (`megatron/training/checkpointing.py`)

- Logits saver is flushed (and its flush future tracked) before checkpoint finalization, ensuring log-prob files are consistent with checkpoint state.
- **`--override-ckpt-iteration`**: overrides the iteration and `consumed_train_samples` stored in a loaded checkpoint, allowing replay of data from an arbitrary iteration (useful for resuming a distillation run at a different position).

---

### Async checkpoint dependencies (`megatron/core/dist_checkpointing/strategies/async_utils.py`)

- Added `finalize_dependencies: List[Future]` field to `AsyncRequest` — finalization functions now wait for all listed futures (e.g. the logits flush future) before running.
- `AsyncCallsQueue._dependencies_complete()` polls dependency futures, with both blocking and non-blocking variants, coordinated across distributed ranks.

---

### Loss function wiring (`pretrain_gpt.py`, `pretrain_hybrid.py`)

- `loss_func` checks for `--logits-load-dir` and, when set, delegates to `LossFuncCallable` (lazy-initialized on first call) instead of the standard or ModelOpt loss.

---

### New arguments (`megatron/training/arguments.py`)

| Argument | Description |
|---|---|
| `--logits-save-top-k` | Number of top logits to save per token |
| `--logits-save-top-p` | Nucleus threshold applied after top-K |
| `--logits-save-top-p-min-k` | Minimum entries kept when top-P is active |
| `--logits-save-dir` | Output directory for teacher log-probs |
| `--logits-save-flush-interval` | Iterations buffered per tar archive |
| `--logits-save-dtype` | On-disk dtype (`fp16` / `bf16` / `fp32`) |
| `--logits-load-dir` | Directory to load cached teacher log-probs |
| `--logits-load-decode-threads` | Worker threads for zstd/torch.load decode |
| `--logits-load-prefetch-factor` | DataLoader prefetch factor |
| `--logits-load-msc-prefetch-depth` | Object-store tar shard prefetch depth |
| `--logits-load-kd-loss-alpha` | Blend weight between KD and LM loss |
| `--logits-load-ignore-errors` | Skip missing/corrupt shards gracefully |
| `--freeze-all-layers` | Freeze all model parameters |
| `--override-ckpt-iteration` | Override loaded checkpoint iteration |

---

### Dependencies (`pyproject.toml`)

- `multi-storage-client`: bumped `~=0.27` → `~=0.49`
- `zstandard`: added (required for tar shard compression/decompression)

File: megatron/core/dist_checkpointing/strategies/async_utils.py [hunk 1]
```diff
@@ -120,6 +123,7 @@ class AsyncRequest(NamedTuple):
     preload_fn: Optional[Callable] = None
     is_frozen: bool = False
     call_idx: int = 0
+    finalize_dependencies: Optional[List[concurrent.futures.Future]] = None
```
Comment (AAnoosheh) [inline megatron/core/dist_checkpointing/strategies/async_utils.py]: NOTE: We may need to upstream this into `nvidia-resiliency-ext` package now that this file is deprecated here, if we don't find an alternate way to non-blockingly finalize the logit flush with the checkpoint finalization.

File: megatron/training/arguments.py [hunk 2]
```diff
@@ -2661,6 +2653,10 @@ def _add_checkpointing_args(parser):
                        help='Do not load optimizer when loading checkpoint.')
     group.add_argument('--no-load-rng', action='store_true', default=None,
                        help='Do not load rng state when loading checkpoint.')
+    group.add_argument('--override-ckpt-iteration', type=int, default=None,
+                       help='Override the iteration stored in the loaded checkpoint. '
+                            'Also resets consumed_train_samples accordingly so the '
+                            'data loader replays samples from that iteration onward.')
```
Comment (AAnoosheh) [inline megatron/training/arguments.py:2677]: NOTE: This feature allows us to begin dumping teacher from any iteration without needing a checkpoint from that iteration.

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/dist_checkpointing/strategies/async_utils.py (hunk 1 above)
Comment (AAnoosheh) [reply megatron/core/dist_checkpointing/strategies/async_utils.py]: No longer needed. We worked around it

File: megatron/training/checkpointing.py [hunk 3]
```diff
@@ -869,7 +870,26 @@ def wandb_finalize_fn():
...
+            async_logits_request = async_request_cls(
+                async_fn=logits_saver._write_batched_tar,
+                async_fn_args=logits_saver.take_pending_data(),
+                finalize_fns=async_save_request.finalize_fns.copy(),
+            )
+            async_save_request.finalize_fns.clear()
+
         schedule_async_save(async_save_request)
+        if logits_saver is not None:
+            schedule_async_save(async_logits_request)
```
[the remark this answers is not in the document]
Comment (AAnoosheh) [reply megatron/training/checkpointing.py:892]: This is on purpose.

**Comment by AAnoosheh** [comment]
/ok to test [d5197e6](https://github.com/NVIDIA/Megatron-LM/pull/5019/commits/d5197e6df7577f14dde6da0cca08f8c410838165)

**Review by deepakn94 — APPROVED** [review]

**Review by aklife97 — APPROVED** [review]

**Review by Victarry — APPROVED** [review]

**Comment by AAnoosheh** [comment]
/ok to test [5fa182c](https://github.com/NVIDIA/Megatron-LM/pull/5019/commits/5fa182c3170b1bfb062550a5217ff7c59e92cdbb)

**Comment by AAnoosheh** [comment]
/ok to test [304ea23](https://github.com/NVIDIA/Megatron-LM/pull/5019/commits/304ea23fd66dc14e202d8d9d89cc96e56bc956d8)
